modules/atmosphere/rendering/atmospheredeferredcaster.cpp

```diff
@@ -1190,369 +1000,565 @@ void AtmosphereDeferredcaster::executeCalculations(GLuint quadCalcVAO,
+        _deltaSProgramObject->setUniform("deltaSMTexture", deltaSMieTableTextureUnit);
+        loadAtmosphereDataIntoShaderProgram(_deltaSProgramObject);
+        glClear(GL_COLOR_BUFFER_BIT);
         for (int layer = 0; layer < static_cast<int>(_r_samples); ++layer) {
-            step3DTexture(_deltaSSupTermsProgramObject, layer, false);
+            step3DTexture(_deltaSProgramObject, layer, false);
             renderQuadForCalc(quadCalcVAO, vertexSize);
         }
         if (_saveCalculationTextures) {
+            saveTextureToPPMFile(GL_COLOR_ATTACHMENT0, std::string("S_texture.ppm"),
+                _mu_s_samples * _nu_samples, _mu_samples);
+        }
+        _deltaSProgramObject->deactivate();
+
+        // loop in line 6 in algorithm 4.1
+        for (int scatteringOrder = 2; scatteringOrder <= 4; ++scatteringOrder) {
+            // line 7 in algorithm 4.1
+            glFramebufferTexture(
+                GL_FRAMEBUFFER,
+                GL_COLOR_ATTACHMENT0,
+                _deltaJTableTexture,
+                0
+            );
+            checkFrameBufferState("_deltaJTableTexture");
+            glViewport(0, 0, _mu_s_samples * _nu_samples, _mu_samples);
+            _deltaJProgramObject->activate();
+            if (scatteringOrder == 2) {
+                _deltaJProgramObject->setUniform("firstIteraction", 1);
+            }
+            else {
+                _deltaJProgramObject->setUniform("firstIteraction", 0);
+            }
+            transmittanceTableTextureUnit.activate();
+            glBindTexture(GL_TEXTURE_2D, _transmittanceTableTexture);
+            _deltaJProgramObject->setUniform(
+                "transmittanceTexture",
+                transmittanceTableTextureUnit
+            );
+            deltaETableTextureUnit.activate();
+            glBindTexture(GL_TEXTURE_2D, _deltaETableTexture);
+            _deltaJProgramObject->setUniform("deltaETexture", deltaETableTextureUnit);
+            deltaSRayleighTableTextureUnit.activate();
+            glBindTexture(GL_TEXTURE_3D, _deltaSRayleighTableTexture);
+            _deltaJProgramObject->setUniform(
+                "deltaSRTexture",
+                deltaSRayleighTableTextureUnit
+            );
+            deltaSMieTableTextureUnit.activate();
+            glBindTexture(GL_TEXTURE_3D, _deltaSMieTableTexture);
+            _deltaJProgramObject->setUniform("deltaSMTexture", deltaSMieTableTextureUnit);
+            loadAtmosphereDataIntoShaderProgram(_deltaJProgramObject);
+            for (int layer = 0; layer < static_cast<int>(_r_samples); ++layer) {
+                step3DTexture(_deltaJProgramObject, layer);
+                renderQuadForCalc(quadCalcVAO, vertexSize);
+            }
+            std::stringstream sst;
+            if (_saveCalculationTextures) {
+                sst << "deltaJ_texture-scattering_order-" << scatteringOrder << ".ppm";
+                saveTextureToPPMFile(GL_COLOR_ATTACHMENT0, sst.str(),
```

Comment:
Better to use fmt::format here:
`fmt::format("deltaJ_texture-scattering_order-{}.ppm", scatteringOrder)`

Reply:
Done.